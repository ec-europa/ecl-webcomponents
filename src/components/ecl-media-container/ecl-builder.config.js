const path = require('path');
const pkg = require('./package.json');

const nodeModules = path.resolve(__dirname, '../../../node_modules');
const outputFolder = './build';

// SCSS includePaths
const includePaths = [nodeModules];

const banner = `${pkg.name} - ${pkg.version} Built on ${new Date().toISOString()}`;

module.exports = {
  scripts: [
    {
      entry: path.resolve(nodeModules, '@ecl/vanilla-component-media-container/media-container.js'),
      dest: path.resolve(outputFolder, 'scripts/ecl-media-container-vanilla.js'),
      options: {
        banner,
        moduleName: 'ECL',
        includePaths,
        sourceMap: false,
      },
    },
  ],
  styles: [
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-media-container-ec.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-media-container-ec.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
        minify: true,
      },
    },
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-media-container-eu.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-media-container-eu.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
        minify: true,
      },
    },
  ],
};
