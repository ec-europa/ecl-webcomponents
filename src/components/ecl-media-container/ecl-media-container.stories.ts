const getArgs = () => {
  return {
    ratio: '16-9',
    fullWidth: false,
    description: 'The European Commission has put forward ambitious yet realistic proposals for a modern EU budget. It is time for an EU budget that reflects rapid developments in innovation, the economy, the environment and geopolitics, amongst others.',
  };
};

const getArgTypes = () => {
  return {
    ratio: {
      control: { name: 'select' },
	    options: {
	      auto: '',
	      '16/9': '16-9',
	      '4/3': '4-3',
	      '3/2': '3-2',
	      '1/1': '1-1',
	    },
	    table: {
	      description: "Ratio of the media"
	    },
	  },
    fullWidth: {
    	control: { name: 'boolean'},
    	table: {
    		description: 'Makes the component render in full width',
    	},
    },
    description: {
    	control: { name: 'string' },
    	table: {
    		description: 'Text to be used as a caption',
    	},
    },
  };
};

export default {
  title: 'Components/media container',
};

const TemplateImg = (args) => 
  `<ecl-media-container
    ratio="${args.ratio}"
    theme="${args.theme}"
  	image="https://inno-ecl.s3.amazonaws.com/media/examples/example-image.jpg"
  >
    ${args.description}
  </ecl-media-container>`;

export const MediaContainerImage = TemplateImg.bind({});
MediaContainerImage.storyName = 'image';
MediaContainerImage.args = getArgs();
MediaContainerImage.argTypes = getArgTypes();

const TemplateVideo = (args) => 
  `<ecl-media-container
    ratio="${args.ratio}"
    theme="${args.theme}"
  	sources='[{"src": "https://inno-ecl.s3.amazonaws.com/media/videos/big_buck_bunny.mp4", "type": "video/mp4"}, {"src": "https://inno-ecl.s3.amazonaws.com/media/videos/big_buck_bunny.webm", "type": "video/webm"}]'
  	tracks='[{"src": "/captions/bunny-en.vtt", "kind": "captions", "src_lang": "en", "label": "English"}, {"src": "/captions/bunny-fr.vtt", "kind": "captions", "src_lang": "fr", "label": "franÃ§ais"}]'
  >
    ${args.description}
  </ecl-media-container>`;

export const MediaContainerVideo = TemplateVideo.bind({});
MediaContainerVideo.storyName = 'video';
MediaContainerVideo.args = getArgs();
MediaContainerVideo.argTypes = getArgTypes();

const TemplateIframe = (args) => 
  `<ecl-media-container
    ratio="${args.ratio}"
    theme="${args.theme}"
    ecl-script
    with-utils
    data-ecl-media-container
  >
    ${args.description}
  	<div class="ecl-media-container__media" slot="embedded-media"><iframe title="New digital strategy" width="350" height="197" src="https://www.youtube.com/embed/fgi-GSCB6ho" frameBorder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowFullScreen></iframe></div>
  </ecl-media-container>`;

export const MediaContainerIframe = TemplateIframe.bind({});
MediaContainerIframe.storyName = 'embedded media';
MediaContainerIframe.args = getArgs();
MediaContainerIframe.argTypes = getArgTypes();







