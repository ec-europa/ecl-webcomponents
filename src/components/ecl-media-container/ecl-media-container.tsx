import { Component, h, Prop, getAssetPath, Element } from '@stencil/core';
declare const ECL: any;

@Component({
  tag: 'ecl-media-container',
  styleUrls: {
    ec: './build/styles/ecl-media-container-ec.css',
    eu: './build/styles/ecl-media-container-eu.css',
  },
  shadow: false,
  scoped: false,
  assetsDirs: ['build'],
})
export class EclMediaContainer {
  @Element() el: HTMLElement;
  @Prop() theme: string = 'ec';
  @Prop() styleClass: string;
  @Prop() alt: string;
  @Prop() description: string;
  @Prop() image: string;
  @Prop() fullWidth: boolean = false;
  @Prop() sources: string;
  @Prop() tracks: string;
  @Prop() ratio: string = '16-9';
  @Prop() eclScript: boolean = false;
  @Prop() withUtils: boolean = false;
  @Prop() embeddedMedia: boolean = false;

  getClass(): string {
    return [
      `ecl-media-container`,
      `ecl-media-container--${this.ratio}`,
      this.styleClass
    ].join(' ');
  }

  componentWillLoad() {
    if (this.withUtils && !document.querySelector('#ecl-utils-styles')) {
      const style = document.createElement('style');
      style.id = 'ecl-utils-styles';
      style.innerHTML = `@import "./build/styles/ecl-utilities-${this.theme}.css"`;
      document.body.appendChild(style);
    }
    if (this.eclScript) {
      const src = getAssetPath('./build/scripts/ecl-media-container-vanilla.js');
      if (document.querySelector(`script[src="${src}"]`)) {
        document.querySelector(`script[src="${src}"]`).remove();
      }
      const script = document.createElement('script');
      script.src = src;
      script.onload = () => {
        const mediaContainer = new ECL.MediaContainer(this.el.firstElementChild);
        mediaContainer.init();
      };
      document.body.appendChild(script);
    }
  }

  render() {
    const sources = this.sources ? JSON.parse(this.sources) : '';
    const tracks = this.tracks ? JSON.parse(this.tracks) : '';
    const alt = this.alt ? { alt: this.alt } : '';
    const videoSources =  sources ? sources.map((source) => (
      <source type={source.type} src={ source.src } />
    )) : '';
    const videoTracks =  tracks ? tracks.map((track) => (
      <track
        kind={track.kind}
        label={track.label}
        src={track.src}
        srclang={track.src_lang}
      />
    )) : '';

    return (
      <figure
        class={this.getClass()}
      >
      { this.image ?
        <img
          class="ecl-media-container__media"
          src={this.image}
          { ...alt }
        />
        : '' }

      { sources && tracks ?
        <video 
          class="ecl-media-container__media"
          poster={this.image}
          controls
        >
          { ...videoSources }
          { ...videoTracks }
        </video> : '' }

        <slot name="embedded-media"></slot>

      { this.description ?
        <figcaption class="ecl-media-container__caption">{this.description}</figcaption>
        : '' }
      </figure> 
    );
  }
}
