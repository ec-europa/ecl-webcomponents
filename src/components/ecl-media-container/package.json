{
  "name": "@ecl/web-components-media-container",
  "author": "European Commission",
  "license": "EUPL-1.2",
  "version": "0.1.0-alpha.0",
  "description": "ECL Media container",
  "main": "ecl-media-container.tsx",
  "private": true,
  "scripts": {
    "build:styles": "cross-env ecl-builder styles",
    "build:scripts": "cross-env ecl-builder scripts",
    "build": "run-s clean build:*",
    "clean": "rimraf build",
    "prepublish": "npm run build"
  },
  "devDependencies": {
    "@ecl/builder": "3.7.1",
    "@ecl/vanilla-component-media-container": "3.7.1",
    "@ecl/vanilla-layout-grid": "3.7.1",
    "@ecl/vanilla-utility-background": "3.7.1",
    "@ecl/vanilla-utility-border": "3.7.1",
    "@ecl/vanilla-utility-clearfix": "3.7.1",
    "@ecl/vanilla-utility-dimension": "3.7.1",
    "@ecl/vanilla-utility-disablescroll": "3.7.1",
    "@ecl/vanilla-utility-display": "3.7.1",
    "@ecl/vanilla-utility-flex": "3.7.1",
    "@ecl/vanilla-utility-float": "3.7.1",
    "@ecl/vanilla-utility-media": "3.7.1",
    "@ecl/vanilla-utility-print": "3.7.1",
    "@ecl/vanilla-utility-screen-reader": "3.7.1",
    "@ecl/vanilla-utility-spacing": "3.7.1",
    "@ecl/vanilla-utility-typography": "3.7.1",
    "@ecl/vanilla-utility-z-index": "3.7.1"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/planctus/ecl-webcomponents.git"
  },
  "bugs": {
    "url": "https://github.com/planctus/ecl-webcomponents/issues"
  },
  "homepage": "https://github.com/planctus/ecl-webcomponents",
  "keywords": [
    "ecl",
    "europa-component-library",
    "design-system",
    "web-components"
  ]
}
