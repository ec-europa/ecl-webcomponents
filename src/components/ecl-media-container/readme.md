# ecl-media-container



<!-- Auto Generated Below -->


## Properties

| Property        | Attribute        | Description | Type      | Default     |
| --------------- | ---------------- | ----------- | --------- | ----------- |
| `alt`           | `alt`            |             | `string`  | `undefined` |
| `description`   | `description`    |             | `string`  | `undefined` |
| `eclScript`     | `ecl-script`     |             | `boolean` | `false`     |
| `embeddedMedia` | `embedded-media` |             | `boolean` | `false`     |
| `fullWidth`     | `full-width`     |             | `boolean` | `false`     |
| `image`         | `image`          |             | `string`  | `undefined` |
| `ratio`         | `ratio`          |             | `string`  | `'16-9'`    |
| `sources`       | `sources`        |             | `string`  | `undefined` |
| `styleClass`    | `style-class`    |             | `string`  | `undefined` |
| `theme`         | `theme`          |             | `string`  | `'ec'`      |
| `tracks`        | `tracks`         |             | `string`  | `undefined` |
| `withUtils`     | `with-utils`     |             | `boolean` | `false`     |


## Dependencies

### Used by

 - [ecl-featured-item](../ecl-featured-item)

### Graph
```mermaid
graph TD;
  ecl-featured-item --> ecl-media-container
  style ecl-media-container fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
