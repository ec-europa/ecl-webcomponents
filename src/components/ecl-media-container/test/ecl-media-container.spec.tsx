import { newSpecPage } from '@stencil/core/testing';
import { EclMediaContainer } from '../ecl-media-container';

describe('ecl-media-container', () => {
  it('renders correctly with an image', async () => {
    const page = await newSpecPage({
      components: [EclMediaContainer],
      html: `<ecl-media-container
              ratio="16-9"
              theme="ec"
              description="caption for the media container"
              image="https://inno-ecl.s3.amazonaws.com/media/examples/example-image.jpg"
            >
            </ecl-media-container>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('renders correctly with an html5 video', async () => {
    const page = await newSpecPage({
      components: [EclMediaContainer],
      html: `<ecl-media-container
              ratio="16-9"
              theme="ec"
              sources='[{"src": "https://inno-ecl.s3.amazonaws.com/media/videos/big_buck_bunny.mp4", "type": "video/mp4"}, {"src": "https://inno-ecl.s3.amazonaws.com/media/videos/big_buck_bunny.webm", "type": "video/webm"}]'
              tracks='[{"src": "/captions/bunny-en.vtt", "kind": "captions", "src_lang": "en", "label": "English"}, {"src": "/captions/bunny-fr.vtt", "kind": "captions", "src_lang": "fr", "label": "franÃ§ais"}]'
            >
              caption for the media container
            </ecl-media-container>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('renders correctly with an embedded video', async () => {
    const page = await newSpecPage({
      components: [EclMediaContainer],
      html: `<ecl-media-container
              ratio="16-9"
              theme="ec"
              ecl-script
              with-utils
              data-ecl-media-container
            >
              caption for the media container
              <div class="ecl-media-container__media" slot="embedded-media"><iframe title="New digital strategy" width="350" height="197" src="https://www.youtube.com/embed/fgi-GSCB6ho" frameBorder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowFullScreen></iframe></div>
            </ecl-media-container>`,
    });

    expect(page.root).toMatchSnapshot();
  });
});
