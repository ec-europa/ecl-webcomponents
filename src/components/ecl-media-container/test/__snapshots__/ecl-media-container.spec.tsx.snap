// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ecl-media-container renders correctly with an embedded video 1`] = `
<ecl-media-container data-ecl-media-container ecl-script="" ratio="16-9" theme="ec">
  <!---->
  caption for the media container
  <figure class="ecl-media-container ecl-media-container--16-9">
    <div class="ecl-media-container__media" slot="embedded-media">
      <iframe allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen="" frameborder="0" height="197" src="https://www.youtube.com/embed/fgi-GSCB6ho" title="New digital strategy" width="350"></iframe>
    </div>
  </figure>
</ecl-media-container>
`;

exports[`ecl-media-container renders correctly with an html5 video 1`] = `
<ecl-media-container ratio="16-9" sources="[{&quot;src&quot;: &quot;https://inno-ecl.s3.amazonaws.com/media/videos/big_buck_bunny.mp4&quot;, &quot;type&quot;: &quot;video/mp4&quot;}, {&quot;src&quot;: &quot;https://inno-ecl.s3.amazonaws.com/media/videos/big_buck_bunny.webm&quot;, &quot;type&quot;: &quot;video/webm&quot;}]" theme="ec" tracks="[{&quot;src&quot;: &quot;/captions/bunny-en.vtt&quot;, &quot;kind&quot;: &quot;captions&quot;, &quot;src_lang&quot;: &quot;en&quot;, &quot;label&quot;: &quot;English&quot;}, {&quot;src&quot;: &quot;/captions/bunny-fr.vtt&quot;, &quot;kind&quot;: &quot;captions&quot;, &quot;src_lang&quot;: &quot;fr&quot;, &quot;label&quot;: &quot;français&quot;}]">
  <!---->
  caption for the media container
  <figure class="ecl-media-container ecl-media-container--16-9">
    <video class="ecl-media-container__media" controls="">
      <source src="https://inno-ecl.s3.amazonaws.com/media/videos/big_buck_bunny.mp4" type="video/mp4">
      <source src="https://inno-ecl.s3.amazonaws.com/media/videos/big_buck_bunny.webm" type="video/webm">
      <track kind="captions" label="English" src="/captions/bunny-en.vtt" srclang="en"></track>
      <track kind="captions" label="français" src="/captions/bunny-fr.vtt" srclang="fr"></track>
    </video>
  </figure>
</ecl-media-container>
`;

exports[`ecl-media-container renders correctly with an image 1`] = `
<ecl-media-container description="caption for the media container" image="https://inno-ecl.s3.amazonaws.com/media/examples/example-image.jpg" ratio="16-9" theme="ec">
  <!---->
  <figure class="ecl-media-container ecl-media-container--16-9">
    <ecl-picture image="https://inno-ecl.s3.amazonaws.com/media/examples/example-image.jpg" imgclass="ecl-media-container__media sc-ecl-media-container-ec" styleclass="ecl-media-container__picture sc-ecl-media-container-ec"></ecl-picture>
  </figure>
</ecl-media-container>
`;
